File: core-tests/src/test/java/overflowdb/testdomains/gratefuldead/SongDb.java
package overflowdb.testdomains.gratefuldead;

import overflowdb.NodeLayoutInformation;
import overflowdb.NodeRef;
import overflowdb.NodeDb;
import scala.Int;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public class SongDb extends NodeDb {
  protected SongDb(NodeRef ref) {
    super(ref);
  }

  private String _name;
  private String _songType;
  private Integer _performances;

  public String name() {
    return _name;
  }

  public String songType() {
    return _songType;
  }

  public Integer performances() {
    return _performances;
  }

  @Override
  public NodeLayoutInformation layoutInformation() {
    return layoutInformation;
  }

  @Override
  public Object property(String key) {
    if (Song.NAME.equals(key)) {
      return _name;
    } else if (key == Song.SONG_TYPE) {
      return _songType;
    } else if (key == Song.PERFORMANCES) {
      return _performances;
    } else {
      return null;
    }
  }

  @Override
  protected void updateSpecificProperty(String key, Object value) {
    if (Song.NAME.equals(key)) {
      this._name = (String) value;
    } else if (Song.SONG_TYPE.equals(key)) {
      this._songType = (String) value;
    } else if (Song.PERFORMANCES.equals(key)) {
      if (value instanceof String) {
        this._performances = Integer.valueOf((String) value);
      } else {
        this._performances = ((Integer) value);
      }
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }


  @Override
  protected void removeSpecificProperty(String key) {
    if (Song.NAME.equals(key)) {
      this._name = null;
    } else if (Song.SONG_TYPE.equals(key)) {
      this._songType = null;
    } else if (Song.PERFORMANCES.equals(key)) {
      this._performances = null;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }

  public static NodeLayoutInformation layoutInformation = new NodeLayoutInformation(
      Song.label,
      new HashSet<>(Arrays.asList(Song.NAME, Song.SONG_TYPE, Song.PERFORMANCES)),
      Arrays.asList(SungBy.layoutInformation, WrittenBy.layoutInformation, FollowedBy.layoutInformation),
      Arrays.asList(FollowedBy.layoutInformation));
}

File: traversal-tests/src/test/scala/overflowdb/traversal/testdomains/gratefuldead/SongDb.scala
package overflowdb.traversal.testdomains.gratefuldead

import overflowdb.traversal._
import overflowdb.{NodeRef, NodeDb}

class SongDb(ref: NodeRef[SongDb]) extends NodeDb(ref) {
  private var _name: String = null
  private var _songType: String = null
  private var _performances: Integer = null

  def name: String = _name
  def songType: String = _songType
  def performances: Integer = _performances

  /* Song --- followedBy --- Song */
  def followedBy: Traversal[Song] = out(FollowedBy.Label).toScalaAs[Song]

  /* Artist <-- sungBy --- Song */
  def sungBy: Traversal[Artist] = out(SungBy.Label).toScalaAs[Artist]

  /* Artist <-- writtenBy --- Song */
  def writtenBy: Traversal[Artist] = out(WrittenBy.Label).toScalaAs[Artist]

  override def property(key: String) =
    key match {
      case Song.PropertyNames.Name => _name
      case Song.PropertyNames.SongType => _songType
      case Song.PropertyNames.Performances => _performances
      case _ => null
    }

  override protected def updateSpecificProperty(key: String, value: Object) =
    key match {
      case Song.PropertyNames.Name =>
        _name = value.asInstanceOf[String]
      case Song.PropertyNames.SongType =>
        _songType = value.asInstanceOf[String]
      case Song.PropertyNames.Performances =>
        _performances = value match {
          case value: String => Integer.valueOf(value)
          case value => value.asInstanceOf[Integer]
        }
      case _ =>
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override protected def removeSpecificProperty(key: String) =
    key match {
      case Song.PropertyNames.Name => _name = null
      case Song.PropertyNames.SongType => _songType = null
      case Song.PropertyNames.Performances => _performances = null
      case _ =>
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override def layoutInformation = Song.layoutInformation
}
